[Refactor] Implement a CoreGraph class according to the design discussion. (2)

daehyun99 opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2383

### List of changes to the codebase in this pull request
- Implement a `TestCoreGraph` class.
- Implement a `CoreGraph` class. (`nx.MultiDiGraph`)

daehyun99 committed:
Refactor: Implement a `CoreGraph` class. (#2383)

- make a list of `CoreGraph`'s method.
- make a list of `TestCoreGraph`'s method.
- make a `CoreGraphABC` class.

daehyun99 committed:
Fix: human error (#2383)

daehyun99 committed:
Refactor: make a frame of `CoreGraph` class (#2383)

- Since `CoreGraphABC` is an internal class, I rename it to `_CoreGraphABC`.
- Added `exposures` and `outcomes` constants to `__init__()`.
- Modified the abstract class. Instead of implementing `is_directed()` and `is_multigraph()` methods in subclasses, we now define `IS_DIRECTED`, `IS_MULTIGRAPH`, and `ALLOWED_EDGES`.

daehyun99 committed:
Temp: Add temp method and constants to implement a `CoreGraph` class. (#2383)

daehyun99 committed:
Test: make a test case list of `CoreGraph` (#2383)

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,248 @@
+from pgmpy.base._mixin_roles import _GraphRolesMixin
+
+
+class _CoreGraphABC(ABC):
```
Wouldn't suggest using `ABC`. It is too strict and creates dependency between the ABC class and the class inheriting it.

ankurankan replied:
I don't think we need this class at all.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,248 @@
+    # And this Constants must be implemented in each graph model that inherits from `CoreGraph`.
+    ALLOWED_EDGES = {"temp"}
+    IS_DIRECTED = False
+    IS_MULTIGRAPH = True
```
I think a simpler way would be to define a single attribute here, say: `SUPPORTED_EDGE_TYPES`. Each inheriting class should define this attribute, and because we don't expect users to use `CoreGraph` directly, we do not need to define attributes like `IS_DIRECTED` or `IS_MULTIGRAPH` because it would be obvious from the object type (i.e., for each of the inheriting classes like `DAG`, `PDAG` we know what kind of graph it would be)

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,248 @@
+        self,
+        ebunch: Optional[Iterable[tuple[Hashable, Hashable]]] = None,
+        exposures: set[Hashable] = set(),
+        outcomes: set[Hashable] = set(),
```
I thought we had decided to not move these arguments out of roles in this PR?

daehyun99 replied:
that work is doing in Pull Request #2400.

I thought #2400 would be merged soon, so I thought it would be good to apply it to `CoreGraph` as well.

What should we do?

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,248 @@
+
+    def is_multigraph(self):
+        """Returns True if graph is a multigraph, False otherwise."""
+        return self.IS_MULTIGRAPH
```
I don't think we will need any of these methods.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,248 @@
+        pass
+
+
+class CoreGraph(_CoreGraphABC, nx.MultiGraph, _GraphRolesMixin):
```
Would suggest renaming this to `_CoreGraph` to signify that it is an internal/private class.

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Feat: Apply code review #2402 (#2383).

- Delete ABC class (`_CoreGraphABC`)
- Delete arguments (`ALLOWED_EDGES`, `IS_DIRECTED`, `IS_MULTIGRAPH`)
- Delete methods (`_graph_type_marker()`, `is_directed()`, `is_multigraph()`)
- Change class name from `CoreGraph` to `_CoreGraph`.

daehyun99 committed:
Test: Write test code for `_CoreGraph.add_edge()` (#2383)

- Write test code for `_CoreGraph.add_edge()`.
- Change name from `CoreGraph` to `_CoreGraph`.

daehyun99 committed:
Test: Remove duplicate test code (#2383).

Check the graph status with the `check_graph_status` function. Duplicate code can be removed.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,191 @@
+
+    """
+
+    SUPPORTED_EDGE_TYPES = ["--", "-*", "*-", "->", "<-", "*>", "<*", "<>", "**"]
```
The `*` needs to be `o`.

daehyun99 committed:
Fix: Apply code review #2402 (#2383)

- Change type expression from `*` to `o`.

daehyun99 committed:
Test: Complete writing test code for `_CoreGraph.add_edge()` (#2383)

daehyun99 committed:
Test: Complete writing test code for `_CoreGraph.__init__()` (#2383)

ankurankan commented on pgmpy/tests/test_base/test_base.py:
```diff
@@ -0,0 +1,312 @@
+from pgmpy.base import _CoreGraph
+
+
+class TestCoreGraph(unittest.TestCase):
```
We are trying to move to pytest from unittest. Could you please modify the tests to not use `unittest`?

ankurankan commented on pgmpy/tests/help_functions.py:
```diff
@@ -3,3 +3,21 @@ def recursive_sorted(li):
     return sorted(li)
+
+
+def check_graph_status(
```
Would suggest keeping this method in the `test_base.py` file itself.

daehyun99 committed:
Feat: Apply code review #2402 (#2383).

- Move position of `check_graph_status()` function.
- Modify the tests to not use `unittest`.

daehyun99 committed:
Test: Complete writing test code for `_CoreGraph.add_edges_from()` (#2383).

- Fix typo error.
- Add test case for missing `type` value.
- Add test case for `u` or `v` is `None`.
- Writing test code for `_CoreGraph.add_edges_from()`.

daehyun99 committed:
Fix: Test code error fix. (#2383)

- Resolve error occurring during the use of `sorted()`.
- Modify the input argument shape of `check_graph_status()`.

daehyun99 committed:
Merge branch 'Refactor/2383/CoreGraph-2' of https://github.com/daehyun99/pgmpy into Refactor/2383/CoreGraph-2

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Feat: Write the code for `add_edge()`, `add_edges_from()` of `_CoreGraph`. (#2383)

- Write the code and comments for the `add_edge()` and `add_edges_from()` methods.
- Created the internal method `_validate_edges_value()` to validate the input values to edge-related methods. I expect this will reduce redundant validation code.
- Separated the locations of external API methods and internal methods with comments.

daehyun99 committed:
Test: Complete writing test code for `remove_edge()`, `remove_edges_from()` methods of `_CoreGraph` (#2383)

- Modify the failing test code for `test_init_with_values`.
- Write the test code for `test_remove_edge`
- Write the test code for `test_remove_edges_from`

daehyun99 committed:
Fix: Fix `ebunch`'s type hint (#2383)

daehyun99 committed:
Fix: Fix value type error in test case. (#2383)

- human error.

daehyun99 committed:
Feat: Write the code for `remove_edge()`, `remove_edges_from()` of _CoreGraph. (#2383)

- Write the code for `remove_edge()`, `remove_edges_from()` of _CoreGraph.
- Change method name from `_validate_edges_value()` -> `_validating_edges_value()`
- Created the internal method `_validating_and_formatting_edges_value()` to validate the input values to edges-related methods. I expect this will reduce redundant validation code.

daehyun99 committed:
Fix: Fix human error. (#2383)

- Modified code for initializing `_CoreGraph`.
- Modified code so that `check_graph_status()` can be executed.
- Prevents the 'KeyError' that can occur when the 'type' value is missing by utilizing 'get()'.

daehyun99 committed:
Test: Complete writing test code for `__eq__()` method of `_CoreGraph`. (#2383)

daehyun99 committed:
Test: Complete writing test code for `copy()` method of `_CoreGraph`. (#2383)

daehyun99 committed:
Fix: Fix human error (#2383)

daehyun99 committed:
Feat: Write the code for `copy()`, `__eq__()` of `_CoreGraph`. (#2383)

- These methods is expected to be usable without being implemented in a subclass of the graph class.

daehyun99 committed:
Docs: Update the cmoonet for `_CoreGraph` (#2383)

daehyun99 committed:
Refactor: Refactor `_CoreGraph`. (#2383)

- Removed `_validating_edges_value()` and replaced it with `_validating_and_formatting_edges_value()`.
- Created the `_get_edge_type_key()` method. A list of tuples in the form of `(u, v, type, key)` can be obtained. It is expected that duplicate code can be removed.

daehyun99 commented:
For your information, I expect the test to pass when the #2400 Pull Request is merged.

daehyun99 committed:
Feat, Docs: Create function to add edges by type. (#2383).

- Create `_is_unknown_edge_type()`, `_is_multigraph()` methods to implement the detailed edge addition algorithm by `type`.
- Create `_add_directed_edge()`, `_add_undirected_edge()`, and `_add_bidirected_edge()` methods subdivided by edge type.
- Unify method comment format

daehyun99 committed:
Docs: Update Comments. (#2383)

daehyun99 committed:
Fix: Inherit from `MultiDiGraph` to support edge directionality. (#2383)

- `_CoreGraph` inheriting from `MultiGraph` can't express directionality in the `add_edge()` method, so I changed it to inherit from `MultiDiGraph`.

daehyun99 committed:
Merge branch 'Refactor/2383/CoreGraph-2' of https://github.com/daehyun99/pgmpy into Refactor/2383/CoreGraph-3

daehyun99 committed:
Feat: Write the code for `add_edge()`. (#2383)

- When writing the code, I changed my mind that `_add_directed_edge()`, `_add_undirected_edge()`, and `_add_bidirected_edge()` are not necessary. 🙄

daehyun99 committed:
Feat: remove the `**kwargs` of `remove_edge()` and `remove_edges_from()` (#2383)

- the `**kwargs` aren't necessary in this method.

daehyun99 committed:
Refactor: Separate `test_init_with_values()` by task. (#2383)

daehyun99 committed:
Refactor: Separate `test_add_edge()` by task. (#2383)

daehyun99 committed:
Refactor: Separate `test_add_edges_from()` by task. (#2383)

daehyun99 committed:
Refactor: Separate `test_remove_edge()` by task. (#2383)

daehyun99 committed:
Refactor: Separate `test_remove_edges_from()` by task. (#2383)

daehyun99 committed:
Refactor: Separate `test_equality()` by task. (#2383)

daehyun99 committed:
Refactor: Separate `test_copy()` by task. (#2383)

daehyun99 committed:
Fix: typo error (#2383)

daehyun99 committed:
Refactor: Separate test code by task. (#2383)

- Refactor: Separate `test_init_with_values()` by task.
- Refactor: Separate `test_add_edge()` by task.
- Refactor: Separate `test_add_edges_from()` by task.
- Refactor: Separate `test_remove_edge()` by task.
- Refactor: Separate `test_remove_edges_from()` by task.
- Refactor: Separate `test_equality()` by task.
- Refactor: Separate `test_copy()` by task.

daehyun99 committed:
Merge branch 'Refactor/2383/CoreGraph-2' of https://github.com/daehyun99/pgmpy into Refactor/2383/CoreGraph-3

daehyun99 committed:
Fix: Correcting error in test code (#2383)

daehyun99 committed:
Feat: Implement the `add_edge()`, `remove_edge()`, and `get_edges()` methods of `_CoreGraph`. (#2383)

- Remove unnecessary code in `remove_edge()` method.
- Change `_is_multigraph()` method to `IS_MULTIGRAPH` value.
- Add an algorithm to find available key values in the `add_edge()` method.
- Since `CoreGraph` inherits from `MultiDiGraph`, the `add_edge()` method has been modified accordingly.
- `CoreGraph` is the most comprehensive graph class capable of representing multiple types of edges. Accordingly, the `get_edges()` method for checking edges has been developed.`
- As `get_edges()` was developed, the test code was also modified.

daehyun99 committed:
Feat: Implement of the `_get_key()` method to find the key value of the edge to be removed in the `remove_edge()` method. (#2383)

- Feat: Implement of the `_get_key()` method to find the key value of the edge to be removed in the `remove_edge()` method.
- Update comments.

daehyun99 committed:
Feat: Implement the `add_edge()`, `remove_edge()`, `get_edges()`, and `_get_key()` methods of `_CoreGraph`. (#2383)

Merge pull request from daehyun99/Refactor/2383/CoreGraph-3 to daehyun99/Refactor/2383/CoreGraph-2.

- Implement of the `_get_key()` method to find the key value of the edge to be removed in the `remove_edge()` method.
- Change `_is_multigraph()` method to `IS_MULTIGRAPH` value.
- Add an algorithm to find available key values in the `add_edge()` method.
- Since `CoreGraph` inherits from `MultiDiGraph`, the `add_edge()` method has been modified accordingly.
- `CoreGraph` is the most comprehensive graph class capable of representing multiple types of edges. Accordingly, the `get_edges()` method for checking edges has been developed.`
- As `get_edges()` was developed, the test code was also modified.

daehyun99 committed:
Refactor: Enhance code readability. && Docs: Update Comments. (#2383)

- Enhance code readability.
- Update Comments.

daehyun99 commented:
Hi, @ankurankan @Spinachboul 
I think all the necessary features and test code for CoreGraph have been developed.
I would appreciate it if you could review the code when you have time.

---

- I defined `IS_MULTIGRAPH` to make the `add_edge()` logic different depending on whether it is a multi-graph.
- If inheriting from `nx.MultiGraph`, edges from u to v and from v to u are added simultaneously. Therefore, I modified it to inherit from `nx.MultiDiGraph`.
- I added the `get_edges()` method to represent both unidirectional edges and bidirectional edges simultaneously.

FYI: https://github.com/pgmpy/pgmpy/pull/2402#issuecomment-3471805178

daehyun99 committed:
Refactor: Add type hints to `_get_key()` method. (#2383

daehyun99 committed:
Refactor: Avoid mutable default arguments in `__init__()` (#2383)

daehyun99 committed:
Fix: Create proper shallow copies of sets in `copy()` method. (#2383)

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+    latents : set of nodes (default: empty `set()`)
+    exposures : set of nodes (default: empty `set()`)
+    outcomes : set of nodes (default: empty `set()`)
+    roles : dict, optional (default: `None`)
```
This needs to numpydoc style: https://numpydoc.readthedocs.io/en/latest/format.html#parameters

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+
+    Examples
+    --------
+"""
```
This is not needed.

ankurankan commented on pgmpy/base/__init__.py:
```diff
@@ -1,7 +1,17 @@
+    "UndirectedGraph",
+    "DAG",
+    "PDAG",
+    "AncestralBase",
```
We wouldn't need `AncestralBase` after `_CoreGraph`. Right?

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+
+class _CoreGraph(nx.MultiDiGraph, _GraphRolesMixin):
+    """
+    Base graph class.
```
I would suggest adding a bit more detail here covering the following points:
1. Generalized representation of all graphs in pgmpy.
2. All graph classes inherit and define their own `SUPPORTED_EDGE_TYPES` to restrict.
3. Provides generalized algorithms that work for all graphs.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+
+    Notes
+    -----
+
```
If any of the docstring sections are empty, you can remove them.

ankurankan commented on pgmpy/base/_base.py:
```diff
@@ -0,0 +1,766 @@
+    Notes
+    -----
+
+    Examples
```
I would suggest keeping the `Examples` section a bit shorter. The idea would be to show some of the main functionalities and then the docstrings in the individual methods can show more functionality/options.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+    {'latents': ['D']}
+    """
+
+    IS_MULTIGRAPH = True
```
What does `IS_MULTIGRAPH` represent? Which of the graph classes among `DAG`, `PDAG`, `MAG`, `PAG`, `ADMG` would it be True for?

daehyun99 replied:
@ankurankan 
> 1. What does `IS_MULTIGRAPH` represent?

I think `_CoreGraph` should be the most general graph from which all graphs in `pgmpy` can inherit.
So I think it should inherit from `networkx.MultiDiGraph`, and the graph's multi-edge capability can be set with `IS_MULTIGRAPH: bool`.

I expect that if a new graph class is developed in `pgmpy` in the future, the class development time can be shortened with `IS_MULTIGRAPH`.

> 2. Which of the graph classes among `DAG`, `PDAG`, `MAG`, `PAG`, `ADMG` would it be True for?

I think they are all `False`.
When writing the method, I also considered the possibility of multi-edges since `ADMG` inherits from `networkx.MultiDiGraph`.

https://github.com/pgmpy/pgmpy/blob/2d10e05dd8f4615edbc7fbe24e8c51a3fe8705d0/pgmpy/base/ADMG.py#L10

Perhaps it inherits from `MultiDiGraph`, considering the situation where `directed edges` and `bidirected edges` exist simultaneously between two nodes.

How about setting the multi-edge status with `IS_MULTIGRAPH` and considering the situation where multiple edges of the same type exist between two nodes with a `new variable`?

ankurankan replied:
In causal graphs, there isn't a scenario where multiple edges of the same type are present between a pair of nodes. Between a given pair of variables, there can be at max 1 edge of 1 type. Hence, I don't think `IS_MULTIGRAPH` would be a useful attribute as it will always be False. I would suggest removing it.

daehyun99 replied:
Yes, I understand.
I will apply the review and remove `IS_MULTIGRAPH`.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+    """
+
+    IS_MULTIGRAPH = True
+    SUPPORTED_EDGE_TYPES = ["--", "-o", "o-", "->", "<-", "o>", "<o", "<>", "oo"]
```
Would be better to store it as a set instead of a list. Would help in faster lookups.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+        --------
+        - Sub-graph classes must implement `SUPPORTED_EDGE_TYPES`
+        - Sub-graph classes must implement `IS_MULTIGRAPH`
+        """
```
I would move this to the `_CoreGraph` docstring instead of here.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+        >>> from pgmpy.base import _CoreGraph
+        >>> G = _CoreGraph()
+        >>> G.SUPPORTED_EDGE_TYPES  # check the available edge types
+        ['--', '-o', 'o-', '->', '<-', 'o>', '<o', '<>', 'oo']
```
I don't think users would directly need to interact with `SUPPORTED_EDGE_TYPES`, so I would suggest avoiding using it in the examples.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+            result.append((u, v, type, key))
+        return result
+
+    def _get_edges_type_key(self):
```
What does `key` represent here?

daehyun99 replied:
@ankurankan 
The [`networkx.MultiDiGraph`](https://github.com/networkx/networkx/blob/d302ea14cb75a80513052ba3a9010f14a203a587/networkx/classes/multidigraph.py#L428)'s `add_edge()` and `remove_edge()` methods have a `key` parameter.

The `key` parameter distinguishes edges between two nodes.

```py
from networkx import MultiDiGraph

G = MultiDiGraph()

G.add_edge("A", "B")
G.add_edge("A", "B")
G.add_edge("A", "B")

G.edges(keys=True)
# OutPut : OutMultiEdgeView([('A', 'B', 0), ('A', 'B', 1), ('A', 'B', 2)])
```

In addition to this, `_CoreGraph` sets the `type` of the edge.

Therefore, when using the `_CoreGraph`'s `add_edge()` and `remove_edge()` methods, `key` allows users to correctly add or remove the desired `type` of edge.

ankurankan replied:
As I mentioned above, there won't be multiple edges of the same type between a pair of variables, so we can remove any functionality related to specifying `key` and let networkx assign default values.

daehyun99 replied:
@ankurankan 
I think the methods related to the `key` value is necessary.

In the `add_edge()` method of `nx.MultiDiGraph`, it's not a problem even if there is no `key` value.
However, in the `remove_edge()` method of [`nx.MultiDiGraph`](https://github.com/networkx/networkx/blob/d302ea14cb75a80513052ba3a9010f14a203a587/networkx/classes/multidigraph.py#L526), it deletes a specific edge using the `u`, `v`, and `key` values.
With the `type` value we defined, we cannot select the edge to delete.

Therefore, I think a method to find the `key` of the edge to be deleted is necessary.

---
Additionally, I have one question regarding the user use case.

There are two graphs where only the edge `key` values are different, and everything else is the same.
In this situation, should `G1.__eq__(G2)` return `True`?
Currently, it is returning `False`.

```py
from pgmpy.base import _CoreGraph

G1 = _CoreGraph()
G1.add_edge("A", "B", "->")
G1.add_edge("A", "B", "<>")
print("G1's edges: ", G1.get_edges(type=True, key=True))
# output: G1's edges:  [('A', 'B', {'type': '->'}, 0), ('A', 'B', {'type': '<>'}, 1)]

G2 = _CoreGraph()
G2.add_edge("A", "B", "<>")
G2.add_edge("A", "B", "->")
print("G2's edges: ", G2.get_edges(type=True, key=True))
# output: G2's edges:  [('A', 'B', {'type': '<>'}, 0), ('A', 'B', {'type': '->'}, 1)]

print(G1.__eq__(G2))
# output: False

```

---
Please let me know if I missed or misunderstood anything.
Thank you.

ankurankan replied:
@daehyun99 Got it, I had missed that `key` would be required for the `remove_edge` method. I am wondering if it would be possible to remove edges based just on the `type` instead of `key`? The only case where we would have multiple edges is in the case of `ADMG` where one would be a directed edge and the other a bidirected edge. If the user provides the type of edge, we can use that to match which edge needs to be removed. What do you think? I feel having both `key` and `type` as arguments in `remove_edge` makes it a bit complex and gets a bit confusing on what each of the arguments is supposed to do.

daehyun99 replied:
>  I am wondering if it would be possible to remove edges based just on the `type` instead of `key`?

Yes, it's possible.
Below is the test code in `test_base.py` that deletes a specific edge only by its `type` value.

```py
    def test_remove_directed_edge(self):
        """Test removing the direct edge of a `_CoreGraph`."""
        edges = [("A", "B", "->"), ("B", "C", "<-")]
        graph = _CoreGraph(ebunch=edges)

        graph.remove_edge("A", "B", "->")
        graph.remove_edge("B", "C", "<-")

        assert not graph.has_edge("A", "B")
        assert not graph.has_edge("B", "C")

        assert graph.get_edges(type=True) == []

        check_graph_status(graph, 3, 0, set(), set(), set(), {})
```

---

> I feel having both `key` and `type` as arguments in `remove_edge` makes it a bit complex and gets a bit confusing on what each of the arguments is supposed to do.

I will modify `remove_edge()` so that the user cannot use the `key` argument.
The `key` value will only be used within the logic.

Thanks.

daehyun99 replied:
FYI: 

> @daehyun99 Got it, I had missed that `key` would be required for the `remove_edge` method. I am wondering if it would be possible to remove edges based just on the `type` instead of `key`? The only case where we would have multiple edges is in the case of `ADMG` where one would be a directed edge and the other a bidirected edge. If the user provides the type of edge, we can use that to match which edge needs to be removed. What do you think? I feel having both `key` and `type` as arguments in `remove_edge` makes it a bit complex and gets a bit confusing on what each of the arguments is supposed to do.

I've modified the code logic to reflect your comments. ([commit](https://github.com/pgmpy/pgmpy/pull/2402/commits/ad0ab4977ce04383e8ba9435b58fe63c8dc9101c))
Now, the `key` value will be declared as the user-input `type` value, instead of an integer.
This means that a maximum of one edge of a specific `type` can exist between two nodes.

Once I finish developing the `get_*` methods, I'll request a review.

Thanks.

Spinachboul commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,766 @@
+            key=lambda x: (x[0], x[1]),
+        )
+
+    def _get_key(
```
Is this function really necessary?
Doesn't networkx handle this implicitly?
`G.edges(keys=True, data=True)`
This should easily return the edges and the keys and the edge's data.
And I dont find much use cases of this helper function elsewhere.
@ankurankan Your thoughts??

Spinachboul commented:
@daehyun99 I would suggest testing the test functions locally before pushing on to the remote branch. This saves us time from getting the same errors while reviewing the PR

daehyun99 commented:
> @daehyun99 I would suggest testing the test functions locally before pushing on to the remote branch. This saves us time from getting the same errors while reviewing the PR

Yes, I understand.

---
@ankurankan @Spinachboul 
I'd like to hear your opinion regarding the implementation of `get_*` methods in `_CoreGraph`.

I'm planning to implement `get_*` (e.g., `get_parents`, `get_children`, ...).
Between the two ways below, which one do you think is better?

---
1. way 1: Finding specific nodes with separate methods (e.g., `get_parents`, `get_children`, `get_spouses`).
I think this is similar to the current implementation of the `AncestralBase` class.
- example
```py
G.get_parents("A") # return parent nodes.
G.get_children("A") # return children nodes.
```

2. way 2: Finding specific nodes in a single method (`get_nodes()`) using a `relation` argument.
- example
```py
G.get_nodes("A", relation="parents") # return parent nodes.
G.get_nodes("A", relation="childrens") # return children nodes.
```

I think way 2(`get_nodes()`) would be better from a user perspective, but I'm worried the logic of a single method might become too complex. 
What do you think?

Thanks.

ankurankan commented:
@daehyun99 I would go with option 1, where the methods are separate. I think that it would be nicer to keep the logic in separate methods; otherwise, we would end up with a long list of if-elses in a single method.

Spinachboul commented:
> @daehyun99 I would go with option 1, where the methods are separate. I think that it would be nicer to keep the logic in separate methods; otherwise, we would end up with a long list of if-elses in a single method.

I agree here since it would be easy and clearer for the users to understand what the function is and what does it do

daehyun99 committed:
Fix: Fix test code for `_CoreGraph` when having multiedges. (#2383)

- There can be at most one edge of a specific `type` between two nodes. Accordingly, the test code is modified.

daehyun99 committed:
Feat, Docs: Apply code review #2402 (#2383)

- Modify the comment's format based on numpydoc.
- Delete `IS_MULTIGRAPH` and related logic.
- Rename and Feat method from `_validating_and_formatting_edges_value` to `_validating_edges_value`.
- Modify the code to hide the `key` value as much as possible, which could confuse users compared `type` value.
- If we match the `key` value and the `type` value, a maximum of one edge of that specific `type` can exist between the two nodes. I have modified the code regarding this:
	- modify logic of `add_edge()`, `remove_edge()`.
	- delete method of `_get_key()`

daehyun99 committed:
Merge branch 'Refactor/2383/CoreGraph-2' of https://github.com/daehyun99/pgmpy into Refactor/2383/CoreGraph-2

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Fix: Fix test code for `_CoreGraph` when having multiedges. (#2383)

- There can be at most one edge of a specific `type` between two nodes. Accordingly, the test code is modified.

daehyun99 commented:
Resolved. ~~This PR's test code depends on #2415.~~

daehyun99 committed:
Fix: Delete part of test code overwriting existing role (#2383)

- It is related #2415.

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,557 @@
+        graph_copy.add_edges_from(ebunch=ebunch)
+        graph_copy.exposures = self.exposures.copy()
+        graph_copy.outcomes = self.outcomes.copy()
+        graph_copy.latents = self.latents.copy()
```
Do we need to manually set these? As all of these attributes are stored as `roles` internally, they should automatically get copied in the next 2 lines?

daehyun99 replied:
> Do we need to manually set these? As all of these attributes are stored as `roles` internally, they should automatically get copied in the next 2 lines?

Yes. I overlooked that.
I think you're right.

daehyun99 committed:
Fix: Apply code review #2402 (#2383)

daehyun99 committed:
Feat: Make frame for `get_*` method. (#2383)

daehyun99 committed:
Feat: Make frame for get_* method more detail. (#2383)

daehyun99 committed:
Test: Write the test code for `get_*` method. (#2383)

- Create `sample_graph1`, `sample_graph2`.
- Copy and Paste from `base_graph` of test_AncestralBase.py to `sample_graph3`.
- Write the test code for `get_*` method.

daehyun99 committed:
Test: Write the failing test code for `get_*` method. (#2383)

daehyun99 committed:
Test: Fix human error. (#2383)

daehyun99 committed:
Test: Fix human error. (#2383)

daehyun99 committed:
Test: Fix typo error. (#2383)

daehyun99 committed:
Test: Implement the test code for `get_*` method of `_CoreGraph` class. (#2383)

> Merge pull request from daehyun99/Refactor/2383/CoreGraph-3 to daehyun99/Refactor/2383/CoreGraph-2.

- Feat: Make frame for `get_* method`.
- Test: Write the test code for `get_* method`.
- Test: Write the failing test code for `get_* method`.

daehyun99 committed:
Test: Fix test case's wrong value. (#2383)

daehyun99 committed:
Test: Add test case for `get_neighbors()` with type values. (#2383)

daehyun99 committed:
Test: Fix test case's wrong value. (#2383)

daehyun99 committed:
Test: Add test case for `get_neighbors()` with type values. (#2383)

> Merge pull request from daehyun99/Refactor/2383/CoreGraph-3 to daehyun99/Refactor/2383/CoreGraph-2.

- Test: Fix test case's wrong value.
- Test: Add test case for `get_neighbors()` with type values.

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Test: Fix Wrong Error type. (#2383)

daehyun99 committed:
Feat: Write the code logic and comments of `get_*` methods of `_CoreGraph`. (#2383)

Docs: Write the comment of `get_*` methods.
Feat: Write the code logic of `get_neighbors()`, `get_parents()`, `get_children()`, `get_spouses()`
Feat: Write code logic of `_validating_nodes_value()` to validating the input of a `get_*`(node-searching) method.

daehyun99 committed:
Test: Add test case for unknown type's edge. (#2383)

daehyun99 committed:
Refactor:  The `get_neighbors()`'s logic of  `_CoreGraph` (#2383)

daehyun99 committed:
Refactor:  The `get_neighbors()`'s logic of  `_CoreGraph` (#2383)

- Improve logic in `get_neighbors` for better efficiency

daehyun99 committed:
Fix: Fix human error (#2383)

daehyun99 committed:
Feat: Write code logic and comments of `get_*` methods of `_CoreGraph`. (#2383)

Merge pull request #23 from daehyun99/Refactor/2383/CoreGraph-3 to daehyun99/Refactor/2383/CoreGraph-2.

- Docs: Write the comment of `get_*` methods.
- Feat: Write the code logic of `get_neighbors()`, `get_parents()`, `get_children()`, `get_spouses()`
- Feat: Write code logic of `_validating_nodes_value()` to validating the input of a `get_*`(node-searching) method.
- Test: Fix Wrong Error type and Add test case for unknown type's edge.

daehyun99 committed:
Fix: Fix test case (#2383)

- `get_ancestors()`, `get_descendants()`, `get_reachable_nodes()` should include node-self.

daehyun99 committed:
Feat: Implements the `get_ancestors()`, `get_descendants()` methods. (#2383)

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,888 @@
+            super().add_edge(v, u, key=type, **kwargs)
+        else:
+            raise AssertionError(
+                "This should never happen."
```
A better way to phrase the error would be something like: `f"Unknown edge type. _CoreGraph currently supports the following edge types: .... Got: {type}")`

daehyun99 committed:
Test: Fix wrong values (#2383)

daehyun99 committed:
Feat: Implements the `get_reachable_nodes()` method (#2383)

daehyun99 committed:
Docs: Apply Review #2402 (#2383)

- Modify the error message.

daehyun99 committed:
Refactor: Refactoring `get_ancestors()`, `get_descendants()`, `get_reachable_nodes()` method. (#2383)

daehyun99 committed:
Refactor: Refactoring `add_edge()`, `remove_edge()` method. (#2383)

daehyun99 committed:
Feat: Implements the `get_*` methods. (#2383)

Merge pull request #24 from daehyun99/Refactor/2383/CoreGraph-3 to daehyun99/Refactor/2383/CoreGraph-2.

- Fix: Fix test case and wrong values :
    - `get_ancestors()`, `get_descendants()`, `get_reachable_nodes()` should include node-self.
- Docs: Apply Review #2402 :
    - Modify the error message.
- Feat: Implements the `get_ancestors()`, `get_descendants()`, `get_reachable_nodes()` methods.
- Refactor: Refactoring `get_ancestors()`, `get_descendants()`, `get_reachable_nodes()` method.
- Refactor: Refactoring `add_edge()`, `remove_edge()` method.

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Docs: Update the comments. (#2383)

daehyun99 commented:
@ankurankan @Spinachboul 
I think that `_CoreGraph` class is ready for merged.
Please review.

Thanks.

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,948 @@
+        self,
+        u: Hashable,
+        v: Hashable,
+        type: str = None,
```
I think we should rename `type` to `edge_type` as `type` is a keyword in Python.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,948 @@
+        # Further adding edge based on the edge's type.
+        if type in ["--", "-o", "o-"]:
+            reverse_type = f"{type[1]}{type[0]}"
+            super().add_edge(v, u, key=reverse_type, **kwargs)
```
Why do we need to add reverse edges?

daehyun99 replied:
> Why do we need to add reverse edges?

@ankurankan 
I will explain this using the `networkx` library code as an example.

In `networkx`, edge information for a graph is stored in `_adj`.

In `pgmpy`, `nx.Graph` is used to represent undirected edges(`UndirectedGraph`), while `nx.DiGraph` is used to represent directed edges(`DAG`).

### In `networkx`
```python
import networkx as nx

graph1 = nx.Graph()
graph1.add_edge("A", "B")
graph1._adj
# output
# {'A': {'B': {}}, 'B': {'A': {}}}
```

The output of `graph1._adj` indicates that edges exist from `A` to `B` and from `B` to `A`.
This represents an undirected edge. (I believe this is how `networkx` internally represents undirected edges.)

```python
import networkx as nx

graph2 = nx.DiGraph()
graph2.add_edge("A", "B")
graph2._adj
# output
# {'A': {'B': {}}, 'B': {}}
```

The output of `graph2._adj` indicates that a directed edge exists from `A` to `B`.
This represents a directed edge.

### In `pgmpy`
```python
from pgmpy.base.base import _CoreGraph

graph3 = _CoreGraph()
graph3.add_edge("A", "B", "->")
graph3._adj
# output
# {'A': {'B': {'->': {}}}, 'B': {}}
```

`_CoreGraph` inherits from `nx.MultiDiGraph`.
It needs to be capable of representing both directed and undirected edges simultaneously.
Therefore, to represent edges that are undirected (`--`) or have unknown directionality (`-o`, `o-`, `oo`), I implemented them using reverse edges.

```python
from pgmpy.base.base import _CoreGraph

graph4 = _CoreGraph()
graph4.add_edge("A", "B", "--")
graph4._adj
# output
# {'A': {'B': {'--': {}}}, 'B': {'A': {'--': {}}}}
```
```python
from pgmpy.base.base import _CoreGraph

graph5 = _CoreGraph()
graph5.add_edge("A", "B", "-o")
graph5._adj
# output
# {'A': {'B': {'-o': {}}}, 'B': {'A': {'o-': {}}}}
```

Please let me know if I have misunderstood anything. Thank you.

ankurankan replied:
@daehyun99 The behavior of the methods that you have implemented is correct. I just think that using two edges to represent certain types of edges complicates the implementation a bit. So, my question is if we only use a single edge for every edge type, can we not make the rest of the methods work? Does that lead to more complicated logic in other methods? Or, could we switch to inheriting from `nx.MultiGraph` instead of `nx.MultiDiGraph` to simplify this?

daehyun99 replied:
@ankurankan 
I'm not entirely sure, but I think inheriting from `nx.MultiGraph` would lead to more complex changes during the refactoring process.

https://github.com/pgmpy/pgmpy/blob/a1a5f2d9767e36e9f83d1f96ae631191d5b5a37d/pgmpy/base/DAG.py#L190-L200

This is because `nx.MultiGraph` doesn't support edge directionality, whereas `DAG`s strictly prohibit cycles. So, I suspect we might not be able to use `nx.find_cycle()`, which is currently utilized in the `DAG._check_cycles()` method.

---

![expect_final](https://github.com/user-attachments/assets/72906cfc-e596-4cea-98ec-cbfd196d6571)

Additionally, I haven't personally worked on the graph classes under `pgmpy/models/`, so I’m not familiar with the implementation details.

However, I suspect there is a high chance of issues arising in classes that inherit from `DAG`, such as `DynamicBayesianNetwork`, `LinearGaussianBayesianNetwork`, and `SEMGraph`.

Therefore, I believe inheriting from `nx.MultiDiGraph` would allow us to handle issues more flexibly during the refactoring process.

---

Furthermore, if the edge handling behavior of `_CoreGraph` and `UndirectedGraph` is consistent within `networkx`, I believe it would lower the complexity of implementing `_GraphAlgorithmMixin`.

What do you think?
Thanks.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,948 @@
+        >>> G.get_edges(type=True)
+        [('A', 'B', {'type': '->'}), ('B', 'C', {'type': '->'})]
+        """
+        self._validating_edges_value(ebunch=ebunch)
```
I don't think we need to call the validation method here. The edges are going to get validated in the `add_edge` method again in line 263.

daehyun99 replied:
> I don't think we need to call the validation method here. The edges are going to get validated in the `add_edge` method again in line 263.

@ankurankan  I agree that the double validation is redundant.

The reason I added line 260 was to ensure that no edges are added to the graph if an error case (like the one below) exists. I thought this all-or-nothing approach would be more user-friendly.

Do you think we should remove line 260?

```python
edges = [("A", "B", "->"), (None, "A", "->"), ("B", "C", "->")]
graph.add_edges_from(ebunch=edges)
```

ankurankan replied:
Yeah, that makes sense. I had missed that.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,948 @@
+        graph_copy.add_edges_from(ebunch=ebunch)
+        graph_copy.exposures = self.exposures
+        graph_copy.outcomes = self.outcomes
+        graph_copy.latents = self.latents
```
Don't think we will need to add these lines as in 401 we are copying the roles which will already have exposure, outcomes, and latents.

ankurankan commented on pgmpy/base/_base.py:
```diff
@@ -0,0 +1,948 @@
+        return sorted(
+            [(u, v, type) for u, v, type in self.edges(keys=True)],
+            key=lambda x: (x[0], x[1]),
+        )
```
I don't think sorting is required for anything other than `__eq__` method. We can also simply use `graphs_equal` method from networkx instead of implementing our own sorting here.

daehyun99 replied:
Yes, I understand.
I removed the sorting logic. 

However, since it's needed for testing, I added a function to `test_base.py` that returns sorted values.(`def get_sorted_edges()`)

Thanks.

ankurankan commented on pgmpy/base/_base.py:
```diff
@@ -0,0 +1,948 @@
+        Returns a list of edges in the graph.
+
+        For undirected and bidirected edges, which are stored as two directed
+        edges internally, this method returns only one of them.
```
I don't think we need to store them as two edges anymore as we are storing the type of edge along with them. This was required earlier in PDAG and ADMG classes as we didn't used to store the edge types.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,948 @@
+            and set(self_ebunch) == set(other_ebunch)
+            and self.exposures == other.exposures
+            and self.outcomes == other.outcomes
+            and self.latents == other.latents
```
Checking for roles dictionary (line 855) should be sufficient, right? These roles would be stored in the roles dictionary as well.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,948 @@
+            and self.get_role_dict() == other.get_role_dict()
+        )
+
+    def _validating_edges_value(
```
Would suggest renaming this to `_validate_edges`.

ankurankan commented on pgmpy/base/base.py:
```diff
@@ -0,0 +1,948 @@
+            key=lambda x: (x[0], x[1]),
+        )
+
+    def _validating_nodes_value(self, node, type):
```
Similary, would suggest renaming this to `_validate_nodes`

ankurankan reviewed (changes requested):
@daehyun99 I have added a few comments. I think it mostly looks good except the part where we are storing multiple edges for undirected and bidirected edges. As we anyways store the edge types now, I don't think we should need to create multiple edges for them. Let me know if I am missing something here.

ankurankan commented on pgmpy/base/_base.py:
Please also rename this file to `_base.py` as this implements non-public class.

ankurankan committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

ankurankan commented:
@daehyun99 I have also fixed some merge conflits on this branch, please make sure to do a `git pull` before pushing anything.

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Feat: Apply Review #2402 && Apply `doctest`'s pre-commit-hock (#2383)

daehyun99 committed:
_CoreGraph is a base graph intended for inheritance, not for direct use.(#2383)

daehyun99 committed:
Docs: Apply doctest (#2383)

daehyun99 committed:
Refactor: Apply Review #2402: Change parameter from `type` to `edge_type`. (#2383)

daehyun99 committed:
Refactor: Apply Review #2402: Change file name from base.py to _base.py (#2383)

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Feat: Apply Review. (#2383)

- Use `nx.utils.graphs_equal()` in `__eq__`

daehyun99 committed:
Feat: Apply Review # 2402 (#2383)

- Delete sort logic
- Add sort logic in Testcode

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

daehyun99 committed:
Test: Modifly test code for apply `_CoreGraph`'s new API (#2402)

daehyun99 committed:
Test: Apply sorting (#2383)

daehyun99 committed:
Test: Modifly test code for apply `_CoreGraph`'s new API (#2383)

daehyun99 committed:
Test: Add test code for `_preprocess_edge` (#2383)

daehyun99 committed:
Test: Fix some test errors (#2383)

daehyun99 committed:
Feat: Apply `_CoreGraph`'s change (#2383)

daehyun99 committed:
Feat: Apply change of `get_neighbors` (#2383)

daehyun99 committed:
Feat, Test: Add code logic and test code for `copy` method's nodes (#2383)

daehyun99 committed:
Feat, Test: Add code logic and test code for `add_edge` method's handling of `kwargs` (#2383)

daehyun99 committed:
Test: Refactor of test code for `_preprocess_edge`. (#2383)

daehyun99 committed:
Feat: Apply change of `_CoreGraph`'s Design. (#2383)

Feat: Apply change of `_CoreGraph`'s Design.
Feat, Test: Add code logic and test code for `copy` method's nodes
Feat, Test: Add code logic and test code for `add_edge` method's handling of `kwargs`
Test: Refactor of test code for _preprocess_edge.

daehyun99 committed:
Refactor: Remove unused method `get_edges` (#2383)

daehyun99 committed:
Merge branch 'Refactor/2383/CoreGraph-2' of https://github.com/daehyun99/pgmpy into Refactor/2383/CoreGraph-3

daehyun99 committed:
Feat: `add_edges_from`

daehyun99 committed:
Test: Add test case for `copy`

daehyun99 committed:
Remove code dup

daehyun99 committed:
Remove code dup

daehyun99 committed:
Refactor: `_validate_edges`

daehyun99 committed:
Refactor: Remove `_validate_nodes`

daehyun99 committed:
Refacotr: Replace `add_edge_from` -> `add_edge`

- `add_edge` will be more fast.

daehyun99 committed:
Test: `remove_edge` policy change

daehyun99 committed:
Refactor: Refactor code logic and test code of `_preprocess_edge`, `_unpreprocess_edge` to `_convert_edge_type`

daehyun99 committed:
Docs: Update comments

daehyun99 committed:
Refactor: Refactor `_CoreGraph` (#2383)

- ENH code readable
- Fix Docs and type hint
- Add test code

daehyun99 commented:
Hi @ankurankan 
I’ve incorporated the feedback from our last meeting.
Please review it at your convenience.

Thanks.

---

I add `_convert_edge_type` method. (`_preprocess_edge` -> `_convert_edge_type` -> `_from_api_edge_type`)
this method translate user's `edge_type` input to store form.
it is replacing `_preprocess_edge`

daehyun99 committed:
Feat: Add `get_edges`, `_to_api_edge_type` && Rename `_convert_edge_type` to `_from_api_edge_type`
Test: Add test code for `get_edges`, `_to_api_edge_type`

daehyun99 committed:
Refactor: Optimize code logic of  `get_edges`

daehyun99 committed:
Refactor: Refactor `_to_api_edge_type`

daehyun99 committed:
Test: Add sorting logic for testing `get_edges`

daehyun99 committed:
Feat: Add `get_edges` (#2383)

- Feat: Add `get_edges`, `_to_api_edge_type` && Rename `_convert_edge_type` to `_from_api_edge_type`
- Test: Add test code for `get_edges`, `_to_api_edge_type`
- Refactor: Optimize code logic of  `get_edges`

daehyun99 committed:
Docs: Update docs part of `get_edges`

daehyun99 committed:
Docs: Update docs part of `get_edges` (#2383)

daehyun99 commented:
![pgmpy edges (16)](https://github.com/user-attachments/assets/61930e60-c6f2-4b7f-bf50-0ff5386ca807)

I add `_from_api_edge_type` and `_to_api_edge_type` to convert edge type's representation. (User API<> pgmpy)

I add `get_edges()` method.
because I think that expression of `get_edges()` is more user friendly than `edges()`.

```py
edges = [("A", "B", "->"), ("A", "B", "<>"), ("B", "C", "->")]
G = _CoreGraph(ebunch=edges)

# get_edges
G.get_edges(keys=True, data=True)
[('A', 'B', 0, '->'), ('A', 'B', 1, '<>'), ('B', 'C', 0, '->')]

# edges
G.edges(keys=True, data=True)
MultiEdgeDataView([('A', 'B', 0, {'A': '-', 'B': '>'}), ('A', 'B', 1, {'A': '>', 'B': '>'}), ('B', 'C', 0, {'B': '-', 'C': '>'})])
```

Please review it at your convenience.

Thanks.

daehyun99 committed:
Docs: Fix human error

ankurankan commented:
@daehyun99 I had a quick look, and this looks good to me on a high level. @Spinachboul If you have time, could you please review this?

daehyun99 committed:
Docs: Update comments (#2383)
Feat: Add `get_edge_type`
Test: Add test code for `get_edge_type`

ankurankan committed:
Merge branch 'dev' into Refactor/2383/CoreGraph-2

ankurankan reviewed (approved):
Looks good to me. Thanks for the massive PR @daehyun99

Spinachboul commented:
@daehyun99 Thanks for PR. Really excited to see how this would turn out.
Hoping for the best!
